app/routes/__auth/user.server.ts

```diff
@@ -14,8 +17,64 @@ export async function getUser({ headers }: Request) {
   const email = session.get('email') as string
   const groups = session.get('groups') as string[]
   const idp = session.get('idp') as string | null
-
+  const refreshToken = session.get('refreshToken') as string
+  const cognitoUserName = session.get('cognitoUserName') as string
+  const accessToken = session.get('accessToken') as string
   if (!sub) return null
+  const user = {
+    sub,
+    email,
+    groups,
+    idp,
+    refreshToken,
+    cognitoUserName,
+    accessToken,
+  }
+  if (!accessToken) await refreshUser(user)
+
+  const tokenSet = new TokenSet(
+    JSON.parse(Buffer.from(accessToken.split('.')[1], 'base64').toString())
+  )
+  if (tokenSet.expired()) await refreshUser(user)
+
+  return user
+}
+
+export async function updateSession(user: any) {
+  const session = await storage.getSession()
+  Object.entries(user).forEach(([key, value]) => {
+    session.set(key, value)
+  })
+  await storage.commitSession(session)
+}
+
+// Clear the access tokens from all sessions belonging to the user with the given sub, to force them to get new access tokens.
+export async function clearUserToken(sub: string) {
+  const db = await tables()
+  const targetSessionResults = await db.sessions.query({
+    IndexName: 'bySub',
+    KeyConditionExpression: '#sub = :sub',
+    ExpressionAttributeNames: {
+      '#sub': 'sub',
+      '#idx': '_idx',
+    },
+    ExpressionAttributeValues: {
+      ':sub': sub,
+    },
+    ProjectionExpression: '#idx',
+  })
+
+  if (!targetSessionResults.Items) return
+
+  const promises = targetSessionResults.Items.map((item) =>
+    db.sessions.update({
+      Key: { _idx: item['_idx'] as string },
+      UpdateExpression: 'REMOVE #token',
+      ExpressionAttributeNames: {
+        '#token': 'token',
+      },
```

Comment:
I think we renamed this field:

```suggestion
      UpdateExpression: 'REMOVE accessToken',
```